src/twisted/web/twcgi.py

```diff
@@ -311,12 +313,24 @@ def processEnded(self, reason):
         if self.errortext:
             self._log.error("Errors from CGI {uri}: {errorText}",
                 uri=self.request.uri, errorText=self.errortext)
+
         if self.handling_headers:
             self._log.error("Premature end of headers in {uri}: {headerText}",
                 uri=self.request.uri, headerText=self.headertext)
-            self.request.write(
-                resource.ErrorPage(http.INTERNAL_SERVER_ERROR,
-                    "CGI Script Error",
-                    "Premature end of script headers.").render(self.request))
-        self.request.unregisterProducer()
-        self.request.finish()
+            if not self._requestFinished:
+                self.request.write(
+                    resource.ErrorPage(http.INTERNAL_SERVER_ERROR,
+                        "CGI Script Error", "Premature end of script headers."
+                    ).render(self.request))
+
+        if not self._requestFinished:
+            self.request.unregisterProducer()
+            self.request.finish()
+
+
+    def _finished(self, ignored):
```

Comment:
I would prefer it if we defined callbacks that are only used once like this as inner methods rather than instance methods to illustrate where they can be used, and to avoid polluting the namespace on an (unfortunately) inheritable class.